test/fuzz/transformation_record_synonymous_constants_test.cpp

```diff
@@ -0,0 +1,286 @@
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint Fragment %4 "main" %22
+               OpExecutionMode %4 OriginUpperLeft
+               OpSource ESSL 310
+               OpName %4 "main"
+               OpName %8 "a"
+               OpName %10 "b"
+               OpName %12 "c"
+               OpName %22 "color"
+               OpDecorate %22 Location 0
+          %2 = OpTypeVoid
+          %3 = OpTypeFunction %2
+          %6 = OpTypeFloat 32
+          %7 = OpTypePointer Function %6
+          %9 = OpConstant %6 0
+         %11 = OpConstantNull %6
+         %13 = OpConstant %6 2
+         %16 = OpTypeBool
+         %20 = OpTypeVector %6 4
+         %21 = OpTypePointer Output %20
+         %22 = OpVariable %21 Output
+         %23 = OpConstantComposite %20 %9 %11 %9 %11
+         %25 = OpConstantComposite %20 %11 %9 %9 %11
+          %4 = OpFunction %2 None %3
+          %5 = OpLabel
+          %8 = OpVariable %7 Function
+         %10 = OpVariable %7 Function
+         %12 = OpVariable %7 Function
+               OpStore %8 %9
+               OpStore %10 %11
+               OpStore %12 %13
+         %14 = OpLoad %6 %8
+         %15 = OpLoad %6 %10
+         %17 = OpFOrdEqual %16 %14 %15
+               OpSelectionMerge %19 None
+               OpBranchConditional %17 %18 %24
+         %18 = OpLabel
+               OpStore %22 %23
+               OpBranch %19
+         %24 = OpLabel
+               OpStore %22 %25
+               OpBranch %19
+         %19 = OpLabel
+               OpReturn
+               OpFunctionEnd
+  )";
+
+  const auto env = SPV_ENV_UNIVERSAL_1_4;
+  const auto consumer = nullptr;
+  const auto context = BuildModule(env, consumer, shader, kFuzzAssembleOption);
+
+  FactManager fact_manager;
+  spvtools::ValidatorOptions validator_options;
+  TransformationContext transformation_context(&fact_manager,
+                                               validator_options);
+  ASSERT_TRUE(IsValid(env, context.get()));
+
+  // %9 and %13 are not equivalent
+  ASSERT_FALSE(TransformationRecordSynonymousConstants(9, 13).IsApplicable(
+      context.get(), transformation_context));
+
+  // %11 and %13 are not equivalent
+  ASSERT_FALSE(TransformationRecordSynonymousConstants(11, 13).IsApplicable(
+      context.get(), transformation_context));
+
+  // %13 and %23 are not equivalent
+  ASSERT_FALSE(TransformationRecordSynonymousConstants(13, 23).IsApplicable(
+      context.get(), transformation_context));
+
+  // %9 and %11 are equivalent
+  ASSERT_TRUE(TransformationRecordSynonymousConstants(9, 11).IsApplicable(
+      context.get(), transformation_context));
+
+  // Apply transformation
+  TransformationRecordSynonymousConstants(9, 11).Apply(context.get(),
+                                                       &transformation_context);
+
+  // Check that the ids are now synonyms
+  transformation_context.GetFactManager()->IsSynonymous(
+      MakeDataDescriptor(9, {}), MakeDataDescriptor(11, {}));
```

Comment:
Please add a case to show that a fact is added when we apply the transformation to two identical float constants.

Reply:
Done